signature/jws/jwt.go

```diff
@@ -25,83 +49,90 @@ func (s *remoteSigningMethod) Verify(signingString, signature string, key interf
 
 // Sign hashes the signingString and call the remote signer to sign the digest.
 func (s *remoteSigningMethod) Sign(signingString string, key interface{}) (string, error) {
-	keySpec, err := s.signer.KeySpec()
-	if err != nil {
-		return "", err
-	}
-
-	// get hasher
-	hasher := keySpec.SignatureAlgorithm().Hash()
-	if !hasher.Available() {
-		return "", &signature.SignatureAlgoNotSupportedError{Alg: hasher.String()}
-	}
-
-	// calculate hash
-	h := hasher.New()
-	h.Write([]byte(signingString))
-	hash := h.Sum(nil)
-
 	// sign by external signer
-	sig, err := s.signer.Sign(hash)
+	sig, certs, err := s.signer.Sign([]byte(signingString))
 	if err != nil {
 		return "", err
 	}
+	s.certs = certs
 	return base64.RawURLEncoding.EncodeToString(sig), nil
 }
 
-// Alg doesn't need to be implemented.
+// Alg return the signing algorithm
 func (s *remoteSigningMethod) Alg() string {
-	alg, err := extractJwtAlgorithm(s.signer)
-	if err != nil {
-		panic(err)
-	}
-	return alg
+	return s.algorithm
 }
 
-// verifyJWT verifies the JWT token against the specified verification key
-func verifyJWT(tokenString string, cert *x509.Certificate) error {
-	keySpec, err := signature.ExtractKeySpec(cert)
-	if err != nil {
-		return err
+// CertificateChain returns the certificate chain
+//
+// should be called after Sign()
+func (s *remoteSigningMethod) CertificateChain() ([]*x509.Certificate, error) {
+	if s.certs == nil {
+		return nil, &signature.RemoteSigningError{Msg: "certificate chain is not set"}
 	}
-	jwsAlg, err := convertAlgorithm(keySpec.SignatureAlgorithm())
+	return s.certs, nil
+}
+
+// PrivateKey returns nil for remote signer
+func (s *remoteSigningMethod) PrivateKey() crypto.PrivateKey {
+	return nil
```

Comment:
Why the remoteSigningMethod needs to implement PrivateKey()?

Reply:
```
type signingMethod interface {
	jwt.SigningMethod

	// CertificateChain returns the certificate chain.
	//
	// should be called after calling Sign()
	CertificateChain() ([]*x509.Certificate, error)

	// PrivateKey returns the private key.
	PrivateKey() crypto.PrivateKey
}
```
`signingMethod interface` includes jwt.SigningMethod called by JWT package when sign the payload, which needs user to provide private key. Both remoteSigningMethod and localSigningMethod implement this interface, then we can eliminate many `if else` cases.